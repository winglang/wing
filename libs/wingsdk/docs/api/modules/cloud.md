[@monadahq/wingsdk](../README.md) / [Exports](../modules.md) / cloud

# Namespace: cloud

## Table of contents

### Enumerations

- [BucketInflightMethods](../enums/cloud.BucketInflightMethods.md)
- [FunctionInflightMethods](../enums/cloud.FunctionInflightMethods.md)
- [QueueInflightMethods](../enums/cloud.QueueInflightMethods.md)

### Classes

- [Bucket](../classes/cloud.Bucket.md)
- [BucketBase](../classes/cloud.BucketBase.md)
- [Function](../classes/cloud.Function.md)
- [FunctionBase](../classes/cloud.FunctionBase.md)
- [Queue](../classes/cloud.Queue.md)
- [QueueBase](../classes/cloud.QueueBase.md)
- [Resource](../classes/cloud.Resource.md)

### Interfaces

- [BucketProps](../interfaces/cloud.BucketProps.md)
- [FunctionProps](../interfaces/cloud.FunctionProps.md)
- [IBucketClient](../interfaces/cloud.IBucketClient.md)
- [IFunctionClient](../interfaces/cloud.IFunctionClient.md)
- [IQueueClient](../interfaces/cloud.IQueueClient.md)
- [QueueOnMessageProps](../interfaces/cloud.QueueOnMessageProps.md)
- [QueueProps](../interfaces/cloud.QueueProps.md)

### Variables

- [BUCKET\_TYPE](cloud.md#bucket_type)
- [FUNCTION\_TYPE](cloud.md#function_type)
- [QUEUE\_TYPE](cloud.md#queue_type)

## Variables

### BUCKET\_TYPE

• `Const` **BUCKET\_TYPE**: ``"wingsdk.cloud.Bucket"``

Global identifier for `Bucket`.

#### Defined in

[src/cloud/bucket.ts:9](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/bucket.ts#L9)

___

### FUNCTION\_TYPE

• `Const` **FUNCTION\_TYPE**: ``"wingsdk.cloud.Function"``

Global identifier for `Function`.

#### Defined in

[src/cloud/function.ts:9](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/function.ts#L9)

___

### QUEUE\_TYPE

• `Const` **QUEUE\_TYPE**: ``"wingsdk.cloud.Queue"``

Global identifier for `Queue`.

#### Defined in

[src/cloud/queue.ts:10](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/queue.ts#L10)
