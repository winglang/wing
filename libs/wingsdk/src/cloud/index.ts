export * from "./api";
export * from "./bucket";
export * from "./counter";
export * from "./domain";
export * from "./endpoint";
export * from "./function";
export * from "./on-deploy";
export * from "./queue";
export * from "./schedule";
export * from "./secret";
export * from "./service";
export * from "./topic";
export * from "./website";
