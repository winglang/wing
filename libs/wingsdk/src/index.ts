// this file must only export the standard library types and not target-specific types.
export * as cloud from "./cloud";
export * as core from "./core";
export * as ex from "./ex";
export * as expect from "./expect";
export * as fs from "./fs";
export * as helpers from "./helpers";
export * as http from "./http";
export * as math from "./math";
export * as platform from "./platform";
export * as regex from "./regex";
export * as aws from "./shared-aws";
export * as simulator from "./simulator";
export * as std from "./std";
export * as sim from "./target-sim";
export * as ui from "./ui";
export * as util from "./util";
