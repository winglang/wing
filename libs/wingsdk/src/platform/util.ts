import { existsSync, readFileSync, readdirSync } from "fs";
import * as path from "path";
import * as toml from "toml";
import * as yaml from "yaml";

/**
 * Creates a JSON object from a comma-separated list of values.
 * Where the key is the path and the value is the value.
 *
 * I.E. "foo/bar=123,foo/baz=456" would result in:
 * {
 * "foo": {
 *  "bar": "123",
 *  "baz": "456"
 *  }
 * }
 *
 * @param values a comma-separated list of values
 * @returns a JSON object with all values
 */
export function parseValuesObjectFromString(values: string) {
  const result: any = {}; // Adjusted type for nested structures

  const valuesList = values.split(",");
  valuesList.forEach((v) => {
    const [paramPath, value] = v.split("=");
    const pathParts = paramPath.split("/");
    let tempObject = result; // Start with the result object

    pathParts.forEach((part, index) => {
      // If we're at the last part, assign the value
      if (index === pathParts.length - 1) {
        tempObject[part] = value;
      } else {
        // Otherwise, initialize the part if it doesn't exist, or use the existing object
        if (!tempObject[part]) tempObject[part] = {};
        tempObject = tempObject[part];
      }
    });
  });

  return result;
}

/**
 * Extracts all fields from a JSON schema.
 *
 * @param schema the schema to extract fields from
 * @returns a set of all fields in the schema
 */
export function extractFieldsFromSchema(schema: any): Set<string> {
  const fields = new Set<string>();

  if (schema.properties) {
    for (const key of Object.keys(schema.properties)) {
      fields.add(key);
    }
  }

  return fields;
}

export function filterParametersBySchema(
  fields: Set<string>,
  parameters: any
): any {
  const filtered: any = {};

  for (const field of fields) {
    if (parameters.hasOwnProperty(field)) {
      filtered[field] = parameters[field];
    }
  }

  return filtered;
}

/**
 * Loads platform-specific values that were passed in via CLI arguments and
 * from a values file. CLI arguments take precedence over values file.
 *
 * I.E. if the cli provided values such as --value "foo/bar=123" and the values file
 * contained:
 *
 * ```yaml
 * foo:
 *  bar: 456
 *  baz: 789
 * ```
 *
 * The resulting values object would be:
 * {
 *  "foo": {
 *    "bar": "123",
 *    "baz": "789"
 *  }
 * }
 *
 * @returns a JSON object with all platform-specific values
 */
export function loadPlatformSpecificValues() {
  const cliValues = parseValuesObjectFromString(process.env.WING_VALUES ?? "");

  if (
    process.env.WING_VALUES_FILE === undefined ||
    process.env.WING_VALUES_FILE === ""
  ) {
    return cliValues;
  }

  const file = path.isAbsolute(process.env.WING_VALUES_FILE)
    ? process.env.WING_VALUES_FILE
    : path.join(process.cwd(), process.env.WING_VALUES_FILE);
  if (!existsSync(file)) {
    return cliValues;
  }
  const data = readFileSync(file, "utf-8");

  const fileExtension = path.extname(file);
  const fileValues = (() => {
    switch (fileExtension) {
      case ".yaml":
      case ".yml":
        return yaml.parse(data);
      case ".json":
        return JSON.parse(data);
      case ".toml":
        return toml.parse(data);
      default:
        throw new Error(
          `Unsupported file extension: ${fileExtension} (expected .yml, .json, or .toml)`
        );
    }
  })();
  return { ...fileValues, ...cliValues };
}

/**
 * Scans a directory for any platform files.
 *
 * @param dir the directory to scan
 * @returns the path to any platform files
 */
export function scanDirForPlatformFile(dir: string): string[] {
  const result: string[] = [];

  if (!existsSync(dir)) {
    return result;
  }

  const files = readdirSync(dir);
  for (const file of files) {
    if (file === "wplatform.js" || file.endsWith(".wplatform.js")) {
      result.push(path.join(dir, file));
    }
  }

  return result;
}
