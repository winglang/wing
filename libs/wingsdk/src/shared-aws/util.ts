export function isValidArn(arn: string, service: string) {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  const { Token } = require("cdktf");

  // if the ARN is an unresolved token, we can't validate it so assume it's valid
  if (Token.isUnresolved(arn)) {
    return true;
  }

  // ARN format: arn:partition:service:region:account-id:resource-type?:resource
  // e.g. arn:aws:lambda:us-east-1:111111111111:function:Function-11111111
  // or,  arn:aws:sqs:us-east-1:111111111111:Queue-11111111
  const parts = arn.split(":");
  if (parts.length < 6 || parts.length > 7) {
    return false;
  }

  if (parts[0] !== "arn") {
    return false;
  }

  if (parts[2] !== service) {
    return false;
  }

  return true;
}
