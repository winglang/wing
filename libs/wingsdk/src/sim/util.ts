import { IConstruct } from "constructs";
import { NodeJsCode } from "../core";
import { Function } from "./function";
import { IResource } from "./resource";

export function captureSimulatorResource(
  type: string,
  resource: IResource,
  captureScope: IConstruct
) {
  if (!(captureScope instanceof Function)) {
    throw new Error(
      `Resources of ${type} can only be captured by a sim.Function for now`
    );
  }

  const env = `${type
    .toUpperCase()
    .replace(/\./g, "_")}_HANDLE_${resource.node.addr.slice(-8)}`;
  const handle = `\${${resource.node.path}#attrs.handle}`; // TODO: proper token mechanism
  captureScope.addEnvironment(env, handle);
  captureScope.node.addDependency(resource);
  resource._addCallers(captureScope.node.path);
  return NodeJsCode.fromInline(
    `$simulator.findInstance(process.env["${env}"])`
  );
}
