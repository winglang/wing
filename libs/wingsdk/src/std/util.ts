/**
 * Generic type argument. This type is replaced at compile time.
 *
 * @hidden
 */
export class T1 {}
