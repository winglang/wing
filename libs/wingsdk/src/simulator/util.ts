import { access, constants } from "node:fs";
import { promisify } from "node:util";

/**
 * Check if a file exists for a specific path
 * @param filePath
 * @Returns Return `true` if the file exists, `false` otherwise.
 */
export async function exists(filePath: string): Promise<boolean> {
  try {
    await promisify(access)(
      filePath,
      constants.F_OK | constants.R_OK | constants.W_OK //eslint-disable-line no-bitwise
    );
    return true;
  } catch (er) {
    return false;
  }
}
