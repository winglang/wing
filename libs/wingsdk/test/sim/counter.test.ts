import * as cloud from "../../src/cloud";
import { ICounterClient } from "../../src/cloud";
import { SimApp, simulatorJsonOf } from "./util";

test("create a counter", async () => {
  // GIVEN
  const app = new SimApp();
  const c = new cloud.Counter(app, "my_counter", {
    initialValue: 123,
  });

  expect(c.initialValue).toBe(123);

  const s = await app.startSimulator();
  expect(s.getAttributes("root/my_counter")).toEqual({
    handle: expect.any(String),
  });
  expect(s.getProps("root/my_counter")).toEqual({
    initialValue: 123,
  });
  await s.stop();

  expect(simulatorJsonOf(s.simfile)).toMatchSnapshot();
});

test("inc", async () => {
  // GIVEN
  const app = new SimApp();
  new cloud.Counter(app, "my_counter", {
    initialValue: 123,
  });

  const s = await app.startSimulator();

  const client = s.getResourceByPath("root/my_counter") as ICounterClient;

  const value0 = await client.inc();
  expect(value0).toEqual(123); // always returns the value before inc (like "i++");

  const value1 = await client.inc();
  expect(value1).toEqual(123 + 1);

  const value2 = await client.inc(10);
  expect(value2).toEqual(123 + 1 + 1);

  const value3 = await client.inc(10);
  expect(value3).toEqual(123 + 1 + 1 + 10);
  await s.stop();

  expect(simulatorJsonOf(s.simfile)).toMatchSnapshot();
});
