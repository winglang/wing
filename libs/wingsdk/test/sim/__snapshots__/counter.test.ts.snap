// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create a counter 1`] = `
Object {
  "root": Object {
    "children": Object {
      "my_counter": Object {
        "attrs": Object {
          "handle": "sim-0",
        },
        "path": "root/my_counter",
        "props": Object {
          "initialValue": 123,
        },
        "type": "wingsdk.cloud.Counter",
      },
    },
    "path": "root",
    "type": "constructs.Construct",
  },
  "sdkVersion": "0.0.0",
  "startOrder": Array [
    "root/my_counter",
  ],
}
`;

exports[`inc 1`] = `
Object {
  "root": Object {
    "children": Object {
      "my_counter": Object {
        "attrs": Object {
          "handle": "sim-0",
        },
        "path": "root/my_counter",
        "props": Object {
          "initialValue": 123,
        },
        "type": "wingsdk.cloud.Counter",
      },
    },
    "path": "root",
    "type": "constructs.Construct",
  },
  "sdkVersion": "0.0.0",
  "startOrder": Array [
    "root/my_counter",
  ],
}
`;
